File: libs/go/candlestick.go
// Candlestick chart renderer — financial OHLC visualization.
// Byte-identical SVG output with the Python renderer.
//
// Proprietary. Copyright (c) 2026 Dharmik Shingala. All rights reserved.
package stonecharts

import (
	"math"
	"strconv"
	"strings"
)

// renderCandlestickSVG delegates to the shared cartesian frame with band
// x-scale, includeZero=false (price axis must not anchor to zero), and
// a11y noun "Candlestick". YAxis.Min/Max are pre-set from OHLC data so the
// frame builds the correct value-axis domain.
func renderCandlestickSVG(spec *ChartSpec) string {
	// Shallow copy to avoid mutating the original.
	s2 := *spec
	series := make([]Series, len(s2.Series))
	copy(series, s2.Series)
	s2.Series = series

	// Pre-set YAxis.Min and YAxis.Max from OHLC data.
	first := true
	var lo, hi float64
	for si := range s2.Series {
		ohlc := s2.Series[si].OHLC
		// Ensure Data has the right length so buildFrame computes n from OHLC length.
		if len(s2.Series[si].Data) == 0 && len(ohlc) > 0 {
			s2.Series[si].Data = make([]float64, len(ohlc))
		}
		for _, d := range ohlc {
			if first {
				lo, hi = d.Low, d.High
				first = false
			}
			if d.Low < lo {
				lo = d.Low
			}
			if d.High > hi {
				hi = d.High
			}
		}
	}
	if s2.YAxis.Min == nil {
		s2.YAxis.Min = &lo
	}
	if s2.YAxis.Max == nil {
		s2.YAxis.Max = &hi
	}

	return renderCartesian(&s2, "Candlestick", "band", candlestickMarks, false)
}

// candlestickMarks emits the candlestick-specific marks — one
// <g class="sc-series"> per series and one candle <g> per datum.
// Subtypes: "" / "candlestick", "ohlc", "hlc", "heikin-ashi", "hollow".
func candlestickMarks(f *cartesianFrame, p *strings.Builder) {
	if f.n <= 0 {
		return
	}

	bandWidth := f.bandWidth()
	groupW := bandWidth * (1 - 0.2)
	k := len(f.spec.Series)
	if k < 1 {
		k = 1
	}
	barW := groupW / float64(k)

	upColor := f.spec.UpColor
	if upColor == "" {
		upColor = "#3f9b6a"
	}
	downColor := f.spec.DownColor
	if downColor == "" {
		downColor = "#d65f5f"
	}

	subtype := f.spec.Subtype

	for si, s := range f.spec.Series {
		p.WriteString(`<g class="sc-series" data-series="` + strconv.Itoa(si) + `">`)

		ohlcData := s.OHLC

		// Heikin-Ashi: transform OHLC before drawing as candlestick.
		if subtype == "heikin-ashi" {
			ha := make([]OHLCDatum, len(ohlcData))
			var prevHaOpen, prevHaClose float64
			for i, d := range ohlcData {
				haClose := (d.Open + d.High + d.Low + d.Close) / 4
				var haOpen float64
				if i == 0 {
					haOpen = (d.Open + d.Close) / 2
				} else {
					haOpen = (prevHaOpen + prevHaClose) / 2
				}
				haHigh := math.Max(d.High, math.Max(haOpen, haClose))
				haLow := math.Min(d.Low, math.Min(haOpen, haClose))
				ha[i] = OHLCDatum{Open: haOpen, High: haHigh, Low: haLow, Close: haClose}
				prevHaOpen = haOpen
				prevHaClose = haClose
			}
			ohlcData = ha
		}

		limit := len(ohlcData)
		if f.n < limit {
			limit = f.n
		}
		for i := 0; i < limit; i++ {
			d := ohlcData[i]
			open, high, low, cl := d.Open, d.High, d.Low, d.Close

			xc := f.xpix(float64(i))
			left := xc - groupW/2 + barW*float64(si)
			cx := left + barW/2
			isUp := cl >= open
			col := downColor
			if isUp {
				col = upColor
			}
			xlabel := strconv.Itoa(i)
			if i < len(f.cats) {
				xlabel = f.cats[i]
			}

			// Candle wrapper <g> with data attributes for interactivity.
			p.WriteString(`<g class="sc-candle sc-point" data-series="` + strconv.Itoa(si) +
				`" data-series-name="` + esc(s.Name) +
				`" data-x="` + esc(xlabel) +
				`" data-y="` + esc(fmtNum(cl)) +
				`" data-open="` + esc(fmtNum(open)) +
				`" data-high="` + esc(fmtNum(high)) +
				`" data-low="` + esc(fmtNum(low)) +
				`" data-close="` + esc(fmtNum(cl)) +
				`" data-color="` + col +
				`" data-r="3.5" data-r-hover="6" cx="` + f1(cx) +
				`" cy="` + f1(f.ypix(cl)) + `">`)

			switch subtype {
			case "ohlc":
				// Vertical line (high-low range).
				p.WriteString(`<line class="sc-wick" x1="` + f1(cx) + `" y1="` + f1(f.ypix(high)) +
					`" x2="` + f1(cx) + `" y2="` + f1(f.ypix(low)) +
					`" stroke="` + col + `" stroke-width="1"/>`)
				// Open tick.
				p.WriteString(`<line class="sc-open-tick" x1="` + f1(left) + `" y1="` + f1(f.ypix(open)) +
					`" x2="` + f1(cx) + `" y2="` + f1(f.ypix(open)) +
					`" stroke="` + col + `" stroke-width="1"/>`)
				// Close tick.
				p.WriteString(`<line class="sc-close-tick" x1="` + f1(cx) + `" y1="` + f1(f.ypix(cl)) +
					`" x2="` + f1(left+barW) + `" y2="` + f1(f.ypix(cl)) +
					`" stroke="` + col + `" stroke-width="1"/>`)

			case "hlc":
				// Vertical line (high-low range).
				p.WriteString(`<line class="sc-wick" x1="` + f1(cx) + `" y1="` + f1(f.ypix(high)) +
					`" x2="` + f1(cx) + `" y2="` + f1(f.ypix(low)) +
					`" stroke="` + col + `" stroke-width="1"/>`)
				// Close tick only (no open tick).
				p.WriteString(`<line class="sc-close-tick" x1="` + f1(cx) + `" y1="` + f1(f.ypix(cl)) +
					`" x2="` + f1(left+barW) + `" y2="` + f1(f.ypix(cl)) +
					`" stroke="` + col + `" stroke-width="1"/>`)

			case "hollow":
				// Wick.
				p.WriteString(`<line class="sc-wick" x1="` + f1(cx) + `" y1="` + f1(f.ypix(high)) +
					`" x2="` + f1(cx) + `" y2="` + f1(f.ypix(low)) +
					`" stroke="` + col + `" stroke-width="1"/>`)
				// Body: up candles get fill="none", down candles get fill=col.
				yTop := f.ypix(math.Max(open, cl))
				yBot := f.ypix(math.Min(open, cl))
				bodyH := math.Max(math.Abs(yBot-yTop), 1.0)
				bodyX := left
				bodyW := barW
				fill := col
				if isUp {
					fill = "none"
				}
				p.WriteString(`<rect class="sc-body" x="` + f1(bodyX) + `" y="` + f1(yTop) +
					`" width="` + f1(bodyW) + `" height="` + f1(bodyH) +
					`" fill="` + fill + `" stroke="` + col + `"/>`)

			default: // "candlestick", "", or "heikin-ashi" (drawn as candlestick)
				// Wick.
				p.WriteString(`<line class="sc-wick" x1="` + f1(cx) + `" y1="` + f1(f.ypix(high)) +
					`" x2="` + f1(cx) + `" y2="` + f1(f.ypix(low)) +
					`" stroke="` + col + `" stroke-width="1"/>`)
				// Body.
				yTop := f.ypix(math.Max(open, cl))
				yBot := f.ypix(math.Min(open, cl))
				bodyH := math.Max(math.Abs(yBot-yTop), 1.0)
				bodyX := left
				bodyW := barW
				p.WriteString(`<rect class="sc-body" x="` + f1(bodyX) + `" y="` + f1(yTop) +
					`" width="` + f1(bodyW) + `" height="` + f1(bodyH) +
					`" fill="` + col + `" stroke="` + col + `"/>`)
			}

			p.WriteString(`</g>`)
		}

		p.WriteString(`</g>`)
	}
}

File: libs/python/stonecharts/charts/candlestick.py
"""Candlestick chart renderer: ChartSpec -> SVG string.

Financial chart rendering for OHLC data. Supports subtypes: candlestick (default),
ohlc, hlc, heikin-ashi, and hollow. Shared Cartesian chrome comes from
_cartesian.py; this module draws only the candle/bar marks.
"""

from __future__ import annotations

import copy

from ..spec import ChartSpec
from ..util import esc, fmt_num
from ._cartesian import CartesianFrame, render_cartesian

PAD = 0.2


def render_svg(spec: ChartSpec) -> str:
    mod = copy.copy(spec)
    mod.y_axis = copy.copy(spec.y_axis)

    # Pre-set y_axis min/max from OHLC highs/lows (only if not already explicit).
    all_lows: list[float] = []
    all_highs: list[float] = []
    for s in mod.series:
        for bar in getattr(s, "ohlc", None) or []:
            all_lows.append(bar["low"])
            all_highs.append(bar["high"])
    if mod.y_axis.min is None and all_lows:
        mod.y_axis.min = min(all_lows)
    if mod.y_axis.max is None and all_highs:
        mod.y_axis.max = max(all_highs)

    for s in mod.series:
        ohlc = getattr(s, "ohlc", None) or []
        if not s.data and ohlc:
            s.data = [0.0] * len(ohlc)

    return render_cartesian(mod, "Candlestick", "band", _candlestick_marks, include_zero=False)


def _candlestick_marks(fr: CartesianFrame, p: list) -> None:
    if fr.n <= 0:
        return

    spec = fr.spec
    up_color = getattr(spec, "up_color", "#3f9b6a")
    down_color = getattr(spec, "down_color", "#d65f5f")
    subtype = getattr(spec, "subtype", "candlestick")

    band_width = fr.band_width()
    group_w = band_width * (1 - PAD)
    k = max(len(spec.series), 1)
    bar_w = group_w / k

    # Pre-compute Heikin-Ashi transformed values if needed.
    ha_data: dict[int, list[dict[str, float]]] = {}
    if subtype == "heikin-ashi":
        for si, s in enumerate(spec.series):
            ohlc = getattr(s, "ohlc", None) or []
            ha_bars: list[dict[str, float]] = []
            prev_ha_open = 0.0
            prev_ha_close = 0.0
            for j, raw in enumerate(ohlc):
                ha_close = (raw["open"] + raw["high"] + raw["low"] + raw["close"]) / 4
                ha_open = (raw["open"] + raw["close"]) / 2 if j == 0 else (prev_ha_open + prev_ha_close) / 2
                ha_high = max(raw["high"], ha_open, ha_close)
                ha_low = min(raw["low"], ha_open, ha_close)
                ha_bars.append(
                    {
                        "open": ha_open,
                        "high": ha_high,
                        "low": ha_low,
                        "close": ha_close,
                    }
                )
                prev_ha_open = ha_open
                prev_ha_close = ha_close
            ha_data[si] = ha_bars

    for si, s in enumerate(spec.series):
        ohlc = getattr(s, "ohlc", None) or []
        p.append(f'<g class="sc-series" data-series="{si}">')

        for i in range(min(len(ohlc), fr.n)):
            bar = ha_data[si][i] if subtype == "heikin-ashi" else ohlc[i]
            o, h, lo, c = bar["open"], bar["high"], bar["low"], bar["close"]

            xc = fr.xpix(i)
            left = xc - group_w / 2 + bar_w * si
            cx = left + bar_w / 2
            is_up = c >= o
            col = up_color if is_up else down_color
            xlabel = fr.cats[i] if i < len(fr.cats) else str(i)

            p.append(
                f'<g class="sc-candle sc-point" data-series="{si}"'
                f' data-series-name="{esc(s.name)}"'
                f' data-x="{esc(xlabel)}" data-y="{esc(fmt_num(c))}"'
                f' data-open="{esc(fmt_num(o))}"'
                f' data-high="{esc(fmt_num(h))}"'
                f' data-low="{esc(fmt_num(lo))}"'
                f' data-close="{esc(fmt_num(c))}"'
                f' data-color="{col}" data-r="3.5" data-r-hover="6"'
                f' cx="{cx:.1f}" cy="{fr.ypix(c):.1f}">'
            )

            if subtype in ("candlestick", "heikin-ashi", "hollow"):
                # Wick line (high to low).
                p.append(
                    f'<line class="sc-wick"'
                    f' x1="{cx:.1f}" y1="{fr.ypix(h):.1f}"'
                    f' x2="{cx:.1f}" y2="{fr.ypix(lo):.1f}"'
                    f' stroke="{col}" stroke-width="1"/>'
                )
                # Body rect.
                y_top = fr.ypix(max(o, c))
                y_bot = fr.ypix(min(o, c))
                body_h = max(abs(y_bot - y_top), 1.0)
                fill = ("none" if is_up else col) if subtype == "hollow" else col
                p.append(
                    f'<rect class="sc-body"'
                    f' x="{left:.1f}" y="{y_top:.1f}"'
                    f' width="{bar_w:.1f}" height="{body_h:.1f}"'
                    f' fill="{fill}" stroke="{col}"/>'
                )
            elif subtype == "ohlc":
                # Vertical line.
                p.append(
                    f'<line class="sc-wick"'
                    f' x1="{cx:.1f}" y1="{fr.ypix(h):.1f}"'
                    f' x2="{cx:.1f}" y2="{fr.ypix(lo):.1f}"'
                    f' stroke="{col}" stroke-width="1"/>'
                )
                # Open tick.
                p.append(
                    f'<line class="sc-open-tick"'
                    f' x1="{left:.1f}" y1="{fr.ypix(o):.1f}"'
                    f' x2="{cx:.1f}" y2="{fr.ypix(o):.1f}"'
                    f' stroke="{col}" stroke-width="1"/>'
                )
                # Close tick.
                p.append(
                    f'<line class="sc-close-tick"'
                    f' x1="{cx:.1f}" y1="{fr.ypix(c):.1f}"'
                    f' x2="{(left + bar_w):.1f}" y2="{fr.ypix(c):.1f}"'
                    f' stroke="{col}" stroke-width="1"/>'
                )
            elif subtype == "hlc":
                # Vertical line.
                p.append(
                    f'<line class="sc-wick"'
                    f' x1="{cx:.1f}" y1="{fr.ypix(h):.1f}"'
                    f' x2="{cx:.1f}" y2="{fr.ypix(lo):.1f}"'
                    f' stroke="{col}" stroke-width="1"/>'
                )
                # Close tick only (no open tick).
                p.append(
                    f'<line class="sc-close-tick"'
                    f' x1="{cx:.1f}" y1="{fr.ypix(c):.1f}"'
                    f' x2="{(left + bar_w):.1f}" y2="{fr.ypix(c):.1f}"'
                    f' stroke="{col}" stroke-width="1"/>'
                )

            p.append("</g>")

        p.append("</g>")
